src/server/worker/master.go

```diff
@@ -811,10 +823,40 @@ func (a *APIServer) waitJob(ctx context.Context, jobInfo *pps.JobInfo, logger *t
 			return a.updateJobState(stm, jobInfo, pps.JobState_JOB_SUCCESS, "")
 		})
 		return nil
-	}, backoff.NewInfiniteBackOff(), func(err error, d time.Duration) error {
+	}, backoff.NewInfiniteBackOff(), func(err error, d time.Duration) (retErr error) {
 		logger.Logf("error in waitJob %v, retrying in %v", err, d)
 		select {
 		case <-ctx.Done():
+			if err := ctx.Err(); err != nil {
+				if err == context.DeadlineExceeded {
+					reason := fmt.Sprintf("job exceeded timeout (%v)", jobInfo.JobTimeout)
+					// Mark the job as failed.
+					// Workers subscribe to etcd for this state change to cancel their work
+					_, err := col.NewSTM(context.Background(), a.etcdClient, func(stm col.STM) error {
+						jobs := a.jobs.ReadWrite(stm)
+						jobID := jobInfo.Job.ID
+						jobInfo := &pps.JobInfo{}
+						if err := jobs.Get(jobID, jobInfo); err != nil {
+							return err
+						}
+						jobInfo.Finished = now()
+						err = a.updateJobState(stm, jobInfo, pps.JobState_JOB_FAILURE, reason)
+						if err != nil {
+							retErr = err
+							return nil
+						}
+						retErr = fmt.Errorf(reason)
+						return nil
+					})
+					if err != nil {
+						return err
+					}
+					if retErr != nil {
+						return retErr
+					}
```

Comment:
I'm very confused what's going on here with `retErr`, I actually don't think you need it at all.

Reply:
Yea, dumb mistake